### PR TITLE
feat: Add provider-neutral repository Git credentials

### PR DESCRIPTION
**Pull request by catkins**
## Description

Repository credentials are currently activated and requested through a provider-specific contract, which prevents the same checkout path from supporting other repository backends. This change introduces a provider-neutral credential contract and uses it for both canonical repositories and optional backend-provided remote Git mirrors.

Remote mirrors remain an optimization: they are attempted only for immutable full commit SHAs, and any miss, timeout, or transport failure falls back to the canonical repository.

This also incorporates the git mirror functionality from https://github.com/buildkite/agent/pull/4132

### Credential flag precedence

Two environment flags can activate managed credentials, and the server dual-emits both for GitHub/GHES during rollout (buildkite/buildkite#31833) so older agents keep working. When both are present the provider-neutral flag wins, otherwise the new path would never be exercised on GitHub:

| `BUILDKITE_USE_REPOSITORY_PROVIDER_GIT_CREDENTIALS` | `BUILDKITE_USE_GITHUB_APP_GIT_CREDENTIALS` | Behavior |
| --- | --- | --- |
| `true` | `true` | Provider-neutral |
| `false` / unset | `true` | Legacy GitHub App |
| `true` | `false` / unset | Provider-neutral |
| `false` / unset | `false` / unset | No managed credentials |

Provider-neutral installs the credential helper before setup, then rewrites SSH to HTTPS and disables SSH keyscan only when the post-environment-hook repository is a `git@github.com:` URL. Legacy installs the helper and applies that rewrite and keyscan change unconditionally before setup. Both paths request tokens from `repository_access_token`, so the precedence rule only affects rewrite and keyscan mode, not the token endpoint.

## Context

The agent needs one credential flow that works across supported repository providers without changing the existing Git credential helper command. Existing users of the legacy client method remain compatible while callers migrate to the generic endpoint.

## Changes

- Activate managed credentials through either the new provider-neutral environment flag or the legacy flag, preferring the provider-neutral path whenever the generic flag is set.
- Restrict provider-neutral SSH-to-HTTPS rewriting and SSH keyscan changes to `git@github.com:` checkouts.
- Add the `repository_access_token` API client and retain the legacy exported method as a deprecated wrapper.
- Update credential helper output, help, and errors to use provider-neutral language.
- Add fail-open exact-SHA checkout support for backend-provided remote Git mirrors, with scoped credential configuration and canonical fallback.
- Protect and allowlist remote mirror configuration alongside the canonical repository.
- Cover credential activation and flag precedence, helper execution, API retries, authenticated mirrors, mirror hits and misses, timeouts, and fallback behavior.

## Testing
- [x] Tests have run locally (with `go test ./...`). Buildkite employees may check this if the pipeline has run automatically.
- [x] Code is formatted (with `go tool gofumpt -extra -w .`)

## Disclosures / Credits

AI assistance was used to implement and test this change under human direction.

### CODE REVIEW

**Review by moskyb — APPROVED** [review]
In principle LGTM from me. i've looked through the code and it all looks pretty much kosher, and i'm all for the concepts of both more generalised git creds, and remote mirrors.

File: internal/job/checkout_remote_mirror.go [hunk 1]
```diff
@@ -0,0 +1,609 @@
...
+		return plan, false, nil
+	}
+
+	var cloneFetchOptions [][]string
+	for i := 0; i < len(gitCloneFlags); i++ {
+		flag := gitCloneFlags[i]
+		switch {
+		case flag == "-v", flag == "--verbose", flag == "-q", flag == "--quiet", flag == "--no-checkout":
+			continue
+		case flag == "--sparse", flag == "--single-branch", flag == "--no-single-branch":
```
[the remark this answers is not in the document]
Comment (pda) [reply internal/job/checkout_remote_mirror.go]: 🤖 Fixed in `8432b0a3`. Remote-mirror plans now persist `remote.origin.tagOpt=--no-tags` and the canonical single-branch `remote.origin.fetch` refspec (including `--depth` unless `--no-single-branch` wins). Single-branch planning resolves only canonical ref metadata with `git ls-remote --symref <canonical> HEAD`; an unresolved HEAD falls back before repository state is created. Differential tests compare canonical and mirror-backed config for all four requested flag combinations and assert the mirror path actually completed.

@buildsworth-bk review

**Comment by pda** [comment]
🤖 Follow-up fix in `8432b0a3` preserves the persistent clone settings that were previously lost on a remote-mirror hit.

- `--no-tags` now persists `remote.origin.tagOpt=--no-tags`.
- Explicit `--single-branch`, and `--depth` unless overridden by `--no-single-branch`, now persist the same narrowed `remote.origin.fetch` refspec as canonical `git clone`. Explicit single-branch flag precedence follows Git’s last-flag-wins behavior.
- The branch for that refspec is resolved from the canonical repository with `git ls-remote --symref <canonical-repo> HEAD`. This transfers only ref metadata; the shallow object transfer still comes from the mirror.
- If canonical HEAD cannot be resolved to `refs/heads/*`, the mirror optimization falls back before creating repository state, allowing canonical `git clone` to retain ownership of the edge-case behavior.

The new differential tests compare `remote.origin.fetch` and `remote.origin.tagOpt` between canonical and mirror-backed clones for `--depth=N`, `--depth=N --no-single-branch`, explicit `--single-branch`, and `--no-tags`. They also verify the test cases completed through the mirror path rather than passing through silent canonical fallback, and cover unresolved-HEAD fallback plus single-branch flag precedence.

**Comment by pda** [comment]
🤖 This combined change has been split into two focused pull requests:

- #4152 contains the provider-neutral repository credentials work, including the rollout fix that preserves legacy pre-setup behavior when both activation flags are present.
- #4153 contains the remote Git mirror checkout optimization and is stacked on #4152 for the provider-neutral credential endpoint.

Closing this pull request as superseded so each concern can be reviewed and landed independently.
